### ISSUE TITLE
Support for "external" configuration of azurerm_private_endpoint/private_dns_zone_group

### ISSUE THREAD

**Issue by arne21a** [terraform-providers/terraform-provider-azurerm#24681]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

tl;dr:
I am looking for a way to control the `private_dns_zone_group` configuration of an `azurerm_private_endpoint` with Azure policy.
The way the provider is currently implemented, the externally created config is overwritten every time and cannot be preserved.  
Therefore, I propose the flag `preserve_existing_dns_config` to exclude this configuration from Terraform.

Long read:
The Azure Cloud Adoption Framework/Enterprise Scale outlines multiple options for private DNS management of private endpoints.
One of them looks like this:

- One central connectivity subscription hosts all private-link hosted zones. This way, the DNS zones can be resolved by multiple spokes.
- Workload subscriptions and their users are able to create private Endpoints within their spoke but do not have permissions to "join" the central private DNS zones.
- Therefore, an Azure policy is deployed to manage the central DNS. It scans for private endpoints and creates the missing DNS config as well as joining them to the central DNS zone.
See: [azure-docs](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/private-link-and-dns-integration-at-scale#private-link-and-dns-integration-in-hub-and-spoke-network-architectures)

The current implementation of `azurerm_private_endpoint` poses a challenge. If the `private_dns_zone_group` block isn't included, it results in the deletion of the existing private_dns_zone_group configuration. Conversely, defining the `private_dns_zone_group` triggers an attempt to enforce a configuration that is beyond the spoke's permission. Consequently, this setup prevents the use of an externally configured DNS management policy when the resource is under Terraform's management.

Proposal:
Add a flag to ignore existing `private_dns_zone_group`.

Workarounds:
I tired a few things to get around this, but none of them are idempotent.  

0. Lifecycle rule - cant be conditional. Not an option

1. Quick and Easy:
- Apply `azurerm_private_endpoint` configuration using terraform
- Apply policy to generate DNS config
- Hardcode DNS config into existing `azurerm_private_endpoint` configuration

2. Quick and less easy:
- Metode No.1, but in terraform with azapi resoruces.
- Apply `azurerm_private_endpoint` configuration using terraform
- Apply policy to generate DNS config
- Enable the `preserve_existing_dns_config` attribute.
- Use azapi to figure out the current config and patch the `azurerm_private_endpoint` on the fly.

This can not be idempotent though, because the terraform dependency tree can not check for a resource with azcli in plan phase.
(there are multiple ways to check, tried a lot of them. Every one had some problem with the dependency tree. Attached my last attempt.)
Original module code can be found [here](https://github.com/aztfmod/terraform-azurerm-caf/tree/main/modules/networking/private_endpoint)

My code:
```
resource "azurecaf_name" "pep" {
  name          = var.name
  resource_type = "azurerm_private_endpoint"
  prefixes      = var.global_settings.prefixes
  random_length = var.global_settings.random_length
  clean_input   = true
  passthrough   = var.global_settings.passthrough
  use_slug      = var.global_settings.use_slug
}

resource "azurerm_private_endpoint" "pep" {
  name                = azurecaf_name.pep.result
  location            = local.location
  resource_group_name = local.resource_group_name
  subnet_id           = var.subnet_id
  tags                = local.tags

  private_service_connection {
    name                           = var.settings.private_service_connection.name
    private_connection_resource_id = var.resource_id
    is_manual_connection           = try(var.settings.private_service_connection.is_manual_connection, false)
    subresource_names              = var.settings.private_service_connection.subresource_names
    request_message                = try(var.settings.private_service_connection.request_message, null)
  }

  dynamic "private_dns_zone_group" {
    for_each = can(var.settings.private_dns) ? [var.settings.private_dns] : []

    content {
      name = lookup(private_dns_zone_group.value, "zone_group_name", "default")
      private_dns_zone_ids = concat(
        flatten([
          for key in try(private_dns_zone_group.value.keys, []) : [
            try(var.private_dns[try(private_dns_zone_group.value.lz_key, var.client_config.landingzone_key)][key].id, [])
          ]
          ]
        ),
        lookup(private_dns_zone_group.value, "ids", [])
      )
    }
  }
  dynamic "private_dns_zone_group" {
    for_each = can(var.settings.preserve_existing_dns_config) && local.pe_exists ? try(local.private_dns_zones_ids) : {}

    content {
      name = "default"
      private_dns_zone_ids = try(private_dns_zone_group.value, [])
    }
  }



  dynamic "ip_configuration" {
    for_each = try(var.settings.ip_configurations, {})

    content {
      name               = ip_configuration.value.name
      private_ip_address = ip_configuration.value.private_ip_address
      subresource_name   = lookup(ip_configuration.value, "subresource_name", null)
      member_name        = lookup(ip_configuration.value, "member_name", null)
    }
  }
}

# get the azurerm_resource_group id from name
data "azurerm_resource_group" "pep_rg" {
  name = local.resource_group_name
}

# get a list of all private endpoints to check if they are already created (kind of redundant, as this cant be read in plan)
data "azapi_resource_list" "private_endpoints" {
  type      = "Microsoft.Network/privateEndpoints@2023-06-01"
  parent_id = format("%s",data.azurerm_resource_group.pep_rg.id)
  response_export_values = ["*"]
}

# get the privateDnsZoneGroups of the current PE
data "azapi_resource_list" "dns_zone_groups" {
  for_each = try(local.pe_exists, false) ? {yes = true} : {}
  type      = "Microsoft.Network/privateEndpoints/privateDnsZoneGroups@2023-06-01"
  parent_id = format("%s/providers/Microsoft.Network/privateEndpoints/%s",data.azurerm_resource_group.pep_rg.id, azurecaf_name.pep.result)
  response_export_values = ["*"]
}

# extract config from azapi data
locals {
  private_dns_zones_ids = try({ for config, data in jsondecode(data.azapi_resource_list.dns_zone_groups["yes"].output).value[0].properties.privateDnsZoneConfigs :
    data.name => [data.properties.privateDnsZoneId]
  }, {})
  available_pe = [for pe in jsondecode(data.azapi_resource_list.private_endpoints.output).value:
    pe.name
  ]
  pe_exists = try(var.settings.preserve_existing_dns_config, false) ? contains(local.available_pe, azurecaf_name.pep.result) : false
}

# some debug output. Use with caution, can break dependencies
resource "null_resource" "debug_output" {
  # Changes to any instance of the cluster requires re-provisioning
  triggers = {
    #raw_api_response = data.azapi_resource_list.dns_zone_groups.output
    raw_api_response_pe = data.azapi_resource_list.private_endpoints.output
    available_pe = try(jsonencode(local.available_pe), [])
    #privateDnsZoneId = jsondecode(data.azapi_resource_list.dns_zone_groups.output).value[0].properties.privateDnsZoneConfigs[0].properties.privateDnsZoneId
    #local = jsonencode(local.private_dns_zones_ids)
    #[0].properties.privateDnsZoneId
  }
}
```

### New or Affected Resource(s)/Data Source(s)

v3.75.0

### Potential Terraform Configuration

```hcl
resource "azurerm_private_endpoint" "pep" {
  name                = azurecaf_name.pep.result
  location            = local.location
  resource_group_name = local.resource_group_name
  subnet_id           = var.subnet_id
  tags                = local.tags
  private_dns_zone_group {
    preserve_existing_dns_config = true
  }
...
}
```

### References
I also found an issue in the caf supermodule project outlining this problem.
https://github.com/aztfmod/terraform-azurerm-caf/issues/559

_No response_

**Comment by mrshiva** [comment]
Met this issue, also, removal of the private endpoint is trying to remove private dns groups.

**Comment by cveld** [comment]
We were discussing this with some coworkers - would be cool if azurerm were able to detect config changes that resulted from DINE (deploy if not exists) policies and just merged them into the terraform state. I guess this requires additional APIs on the Azure control plane side first like "Hey dear resource can you tell me which DINE policies will be / have been applied to you". On the other hand, is Azure Policy actually the best way to define dns config. Or maybe Azure Policy and resource deployment should be an atomically integrated experience. Anyways, I guess the azurerm provider can't do much about this currently in general.
